// @tsed/logger/lib/core/LogEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class LogEvent {
    /**
     * Models a logging event.
     * @constructor
     * @param {String} _categoryName name of category
     * @param {LogLevel} _level level of message
     * @param {Array} _data objects to log
     * @param _context
     */
    constructor(_categoryName, _level, _data, _context) {
        this._categoryName = _categoryName;
        this._level = _level;
        this._data = _data;
        this._context = _context;
        this._startTime = new Date();
    }
    get startTime() {
        return this.data && this.data[0] && this.data[0].time ? this.data[0].time : this._startTime;
    }
    get categoryName() {
        return this._categoryName;
    }
    get level() {
        return this._level;
    }
    get formatedLevel() {
        return (this.level.toString() + "     ").slice(0, 5);
    }
    get data() {
        return this._data;
    }
    get context() {
        return this._context;
    }
    get cluster() {
        return {};
    }
    get pid() {
        return this.context.get("pid");
    }
}
exports.LogEvent = LogEvent;
