// @taquito/michel-codec/dist/lib/utils.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkTezosID = exports.tezosPrefix = exports.unpackAnnotations = exports.isNatural = exports.isDecimal = exports.compareBytes = exports.parseBytes = exports.LongInteger = exports.isMichelsonError = exports.MichelsonError = void 0;
var base58_1 = require("./base58");
var MichelsonError = /** @class */ (function (_super) {
    __extends(MichelsonError, _super);
    /**
     * @param val Value of a AST node caused the error
     * @param path Path to a node caused the error
     * @param message An error message
     */
    function MichelsonError(val, message) {
        var _this = _super.call(this, message) || this;
        _this.val = val;
        Object.setPrototypeOf(_this, MichelsonError.prototype);
        return _this;
    }
    return MichelsonError;
}(Error));
exports.MichelsonError = MichelsonError;
function isMichelsonError(err) {
    return err instanceof MichelsonError;
}
exports.isMichelsonError = isMichelsonError;
// Ad hoc big integer parser
var LongInteger = /** @class */ (function () {
    function LongInteger(arg) {
        this.neg = false;
        this.buf = [];
        if (arg === undefined) {
            return;
        }
        if (typeof arg === "string") {
            for (var i = 0; i < arg.length; i++) {
                var c = arg.charCodeAt(i);
                if (i === 0 && c === 0x2d) {
                    this.neg = true;
                }
                else {
                    if (c < 0x30 || c > 0x39) {
                        throw new Error("unexpected character in integer constant: " + arg[i]);
                    }
                    this.append(c - 0x30);
                }
            }
        }
        else if (arg < 0) {
            this.neg = true;
            this.append(-arg);
        }
        else {
            this.append(arg);
        }
    }
    LongInteger.prototype.append = function (c) {
        var i = 0;
        while (c !== 0 || i < this.buf.length) {
            var m = (this.buf[i] || 0) * 10 + c;
            this.buf[i++] = m % 256;
            c = Math.floor(m / 256);
        }
    };
    LongInteger.prototype.cmp = function (arg) {
        if (this.neg !== arg.neg) {
            return (arg.neg ? 1 : 0) - (this.neg ? 1 : 0);
        }
        else {
            var ret = 0;
            if (this.buf.length !== arg.buf.length) {
                ret = this.buf.length < arg.buf.length ? -1 : 1;
            }
            else if (this.buf.length !== 0) {
                var i = arg.buf.length - 1;
                while (i >= 0 && this.buf[i] === arg.buf[i]) {
                    i--;
                }
                ret = i < 0 ? 0 : this.buf[i] < arg.buf[i] ? -1 : 1;
            }
            return !this.neg ? ret : ret === 0 ? 0 : -ret;
        }
    };
    Object.defineProperty(LongInteger.prototype, "sign", {
        get: function () {
            return this.buf.length === 0 ? 0 : this.neg ? -1 : 1;
        },
        enumerable: false,
        configurable: true
    });
    return LongInteger;
}());
exports.LongInteger = LongInteger;
function parseBytes(s) {
    var ret = [];
    for (var i = 0; i < s.length; i += 2) {
        var x = parseInt(s.slice(i, i + 2), 16);
        if (Number.isNaN(x)) {
            return null;
        }
        ret.push(x);
    }
    return ret;
}
exports.parseBytes = parseBytes;
function compareBytes(a, b) {
    if (a.length !== b.length) {
        return a.length < b.length ? -1 : 1;
    }
    else if (a.length !== 0) {
        var i = 0;
        while (i < a.length && a[i] === b[i]) {
            i++;
        }
        return i === a.length ? 0 : a[i] < b[i] ? -1 : 1;
    }
    else {
        return 0;
    }
}
exports.compareBytes = compareBytes;
function isDecimal(x) {
    try {
        // tslint:disable-next-line: no-unused-expression
        new LongInteger(x);
        return true;
    }
    catch (_a) {
        return false;
    }
}
exports.isDecimal = isDecimal;
function isNatural(x) {
    try {
        return new LongInteger(x).sign >= 0;
    }
    catch (_a) {
        return false;
    }
}
exports.isNatural = isNatural;
var annRe = /^(@%|@%%|%@|[@:%]([_0-9a-zA-Z][_0-9a-zA-Z\.%@]*)?)$/;
function unpackAnnotations(p, opt) {
    var e_1, _a;
    var field;
    var type;
    var vars;
    if (p.annots !== undefined) {
        try {
            for (var _b = __values(p.annots), _c = _b.next(); !_c.done; _c = _b.next()) {
                var v = _c.value;
                if (v.length !== 0) {
                    if (!annRe.test(v) ||
                        (!(opt === null || opt === void 0 ? void 0 : opt.specialVar) && (v === "@%" || v === "@%%")) ||
                        (!(opt === null || opt === void 0 ? void 0 : opt.specialFields) && v === "%@")) {
                        throw new MichelsonError(p, p.prim + ": unexpected annotation: " + v);
                    }
                    switch (v[0]) {
                        case "%":
                            if ((opt === null || opt === void 0 ? void 0 : opt.emptyFields) || v.length > 1) {
                                field = field || [];
                                field.push(v);
                            }
                            break;
                        case ":":
                            if (v.length > 1) {
                                type = type || [];
                                type.push(v);
                            }
                            break;
                        case "@":
                            if ((opt === null || opt === void 0 ? void 0 : opt.emptyVar) || v.length > 1) {
                                vars = vars || [];
                                vars.push(v);
                            }
                            break;
                    }
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
    }
    return { f: field, t: type, v: vars };
}
exports.unpackAnnotations = unpackAnnotations;
exports.tezosPrefix = {
    BlockHash: [32, [1, 52]],
    OperationHash: [32, [5, 116]],
    OperationListHash: [32, [133, 233]],
    OperationListListHash: [32, [29, 159, 109]],
    ProtocolHash: [32, [2, 170]],
    ContextHash: [32, [79, 199]],
    ED25519PublicKeyHash: [20, [6, 161, 159]],
    SECP256K1PublicKeyHash: [20, [6, 161, 161]],
    P256PublicKeyHash: [20, [6, 161, 164]],
    ContractHash: [20, [2, 90, 121]],
    CryptoboxPublicKeyHash: [16, [153, 103]],
    ED25519Seed: [32, [13, 15, 58, 7]],
    ED25519PublicKey: [32, [13, 15, 37, 217]],
    SECP256K1SecretKey: [32, [17, 162, 224, 201]],
    P256SecretKey: [32, [16, 81, 238, 189]],
    ED25519EncryptedSeed: [56, [7, 90, 60, 179, 41]],
    SECP256K1EncryptedSecretKey: [56, [9, 237, 241, 174, 150]],
    P256EncryptedSecretKey: [56, [9, 48, 57, 115, 171]],
    SECP256K1PublicKey: [33, [3, 254, 226, 86]],
    P256PublicKey: [33, [3, 178, 139, 127]],
    SECP256K1Scalar: [33, [38, 248, 136]],
    SECP256K1Element: [33, [5, 92, 0]],
    ED25519SecretKey: [64, [43, 246, 78, 7]],
    ED25519Signature: [64, [9, 245, 205, 134, 18]],
    SECP256K1Signature: [64, [13, 115, 101, 19, 63]],
    P256Signature: [64, [54, 240, 44, 52]],
    GenericSignature: [64, [4, 130, 43]],
    ChainID: [4, [87, 82, 0]],
};
function checkTezosID(id) {
    var e_2, _a;
    var types = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        types[_i - 1] = arguments[_i];
    }
    var buf = Array.isArray(id) ? id : base58_1.decodeBase58Check(id);
    try {
        for (var types_1 = __values(types), types_1_1 = types_1.next(); !types_1_1.done; types_1_1 = types_1.next()) {
            var t = types_1_1.value;
            var prefix = exports.tezosPrefix[t];
            if (buf.length === prefix[0] + prefix[1].length) {
                var i = 0;
                while (i < prefix[1].length && buf[i] === prefix[1][i]) {
                    i++;
                }
                if (i === prefix[1].length) {
                    return [t, buf.slice(prefix[1].length)];
                }
            }
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (types_1_1 && !types_1_1.done && (_a = types_1.return)) _a.call(types_1);
        }
        finally { if (e_2) throw e_2.error; }
    }
    return null;
}
exports.checkTezosID = checkTezosID;
