// rx-sandbox/dist/src/marbles/parseSubscriptionMarble.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var SubscriptionLog_1 = require("rxjs/internal/testing/SubscriptionLog");
var ObservableMarbleToken_1 = require("./ObservableMarbleToken");
var tokenParseReducer_1 = require("./tokenParseReducer");
var parseSubscriptionMarble = function (marble, frameTimeFactor, maxFrame) {
    if (frameTimeFactor === void 0) { frameTimeFactor = 1; }
    if (maxFrame === void 0) { maxFrame = 1000; }
    if (!marble) {
        return new SubscriptionLog_1.SubscriptionLog(Number.POSITIVE_INFINITY);
    }
    var marbleTokenArray = Array.from(marble).filter(function (token) { return token !== ObservableMarbleToken_1.ObservableMarbleToken.NOOP; });
    var value = marbleTokenArray.reduce(tokenParseReducer_1.subscriptionTokenParseReducer(frameTimeFactor, maxFrame), {
        currentTimeFrame: 0,
        subscriptionFrame: Number.POSITIVE_INFINITY,
        unsubscriptionFrame: Number.POSITIVE_INFINITY,
        simultaneousGrouped: false,
        expandingTokenCount: 0,
        expandingValue: []
    });
    return value.unsubscriptionFrame === Number.POSITIVE_INFINITY
        ? new SubscriptionLog_1.SubscriptionLog(value.subscriptionFrame)
        : new SubscriptionLog_1.SubscriptionLog(value.subscriptionFrame, value.unsubscriptionFrame);
};
exports.parseSubscriptionMarble = parseSubscriptionMarble;
